src/software/backend/replay_backend.cpp

```diff
@@ -0,0 +1,77 @@
+#include "replay_backend.h"
+
+#include <cstdlib>
+
+#include "software/util/design_patterns/generic_factory.h"
+
+ReplayBackend::ReplayBackend(const std::string& replay_input_dir)
+    : Backend(),
+      replay_reader(replay_input_dir),
+      pull_from_replay_thread(
+          boost::bind(&ReplayBackend::continuouslyPullFromReplayFiles, this)),
+      last_msg_received_time(std::nullopt),
+      last_msg_replayed_time(std::nullopt),
+      last_primitive_received_time(std::nullopt),
+      last_primitive_received_time_mutex()
+{
+}
+
+// TODO: https://github.com/UBC-Thunderbots/Software/issues/1745
+//       perhaps we want to encode the output primitives to proto and log them?
+void ReplayBackend::onValueReceived(TbotsProto::PrimitiveSet primitives)
+{
+    // update the time when the backend received the last primitive message. this is
+    // used to check when we should exit.
+    std::scoped_lock lock(last_primitive_received_time_mutex);
+    last_primitive_received_time = std::chrono::steady_clock::now();
+}
+
+// do nothing
+void ReplayBackend::onValueReceived(World world) {}
+
+void ReplayBackend::continuouslyPullFromReplayFiles()
+{
+    while (auto sensor_msg_or_null = replay_reader.getNextMsg())
+    {
+        auto this_msg_received_time = std::chrono::duration<double>(
+            sensor_msg_or_null->time_received().epoch_timestamp_seconds());
+
+        if (last_msg_received_time && last_msg_replayed_time)
+        {
+            std::this_thread::sleep_until(
+                *last_msg_replayed_time +
+                (this_msg_received_time - *last_msg_received_time));
+        }
```

Comment:
can we add a quick comment to explain that we want to replicate the timing of messages by sleeping